Delay validation of registered Active Record adapters and allow aliasing of adapters

kmcphillips opened the pull request:
Next step in https://github.com/rails/rails/pull/50064 by @byroot.

## Problem

We want to allow aliasing of DB adapters by name. The end state here is to allow `mysql` to be mapped to either `trilogy` or `mysql2` at an application level. And allow others to do the same.

The main refactor here is moving the adapter validation out of `DatabaseConfig.new` and into a `#validate` method that is called later.

The reason for this refactor is that currently database configs are loaded by the framework so early that there isn't really a way to hook into registering/aliasing adapters before the configs are loaded and validated with `.new`:

https://github.com/rails/rails/blob/438cad462638b02210fc48b700c29dcd0428a8b7/activerecord/lib/active_record/railtie.rb#L305-L307

The adapter needs to be registered with Active Record after it is loaded, but before configs are loaded, and because `on_load(:active_record)` hooks are executed in order the above one is always executed before anything in the rails app, including the initializers. I even tried adding a `:active_record_register_adapters` hook, but again the app initializers aren't run before this fires.


## Solution

As @matthewd suggested, extract validation into a method on `DatabaseConfig`. This has the added bonus of allowing the config subclasses further specialize validation.

Also adds a method which I will use in the next step to `alias("trilogy", as: "mysql")`.


### Reproduction steps

I have validated this in a new rails app, by naming the "sqlite3" adapter "potato", and the app will run with the following steps:

```
$ ../rails/railties/exe/rails new example --dev --skip-asset-pipeline --skip-javascript --skip-hotwire --skip-test
```
```
$ sed -i 's/adapter: sqlite3/adapter: potato/' config/database.yml
```
```
$ bin/rails environment
bin/rails aborted!
ActiveRecord::AdapterNotFound: database configuration specifies nonexistent 'potato' adapter. Ensure that the adapter is spelled correctly in config/database.yml and that you've added the necessary adapter gem to your Gemfile. (ActiveRecord::AdapterNotFound)
```
```
$ echo 'ActiveSupport.on_load(:active_record) { ActiveRecord::ConnectionAdapters.register("potato", "ActiveRecord::ConnectionAdapters::SQLite3Adapter", "active_record/connection_adapters/sqlite3_adapter") }' > config/initializers/active_record_adapters.rb
```
```
$ bin/rails environment
$ bin/rails runner 'puts "#{ ActiveRecord::Base.connection.pool.db_config.adapter } / #{ ActiveRecord::Base.connection.adapter_name }"'
potato / SQLite
```

kmcphillips committed:
Delay validation of registered Active Record adapters. Allow adapter aliasing.

kmcphillips commented on activerecord/lib/active_record/database_configurations/database_config.rb:
```diff
@@ -28,6 +22,12 @@ def new_connection
         adapter_class.new(configuration_hash)
       end
 
+      def validate
```
Naming is hard. This could be `validate!` since it raises on invalid, rather than returning false.

matthewd commented:
Is aliasing here really common enough to warrant specific API? For a pretty specialist requirement, `ActiveRecord::ConnectionAdapters.register("mysql", "ActiveRecord::ConnectionAdapters::TrilogyAdapter")` seems fine to me?

kmcphillips commented:
Yeah I went back and forth on that. I'll remove it, it's fine. It's just a matter of documenting exactly what is expected as params there. 

I'll simplify this diff then. And get CI passing.

kmcphillips commented on activerecord/lib/active_record/connection_adapters/abstract/connection_handler.rb:
```diff
@@ -322,7 +322,7 @@ def disconnect_pool_from_pool_manager(pool_manager, role, shard)
         def resolve_pool_config(config, connection_name, role, shard)
           db_config = Base.configurations.resolve(config)
-
+          db_config.validate
```
CI caught this. The `ridiculous` adapter from another test was passed into `#establish_connection` and a pool was created with this config, but then every time it was attempted to be used it raised because the adapter was invalid.

Here, since we delay validating the connection to an initializer, and not when the class is initialized, it's possible to have invalid connection configs. I bet we have other bugs of this class we haven't caught yet.

matthewd replied:
Is this _sufficient_? (As in, without the initializer?)

I like the idea of proactively validating things while we're setting up the pool, rather than only noticing a problem when we try to use the config to actually connect... but while the initializer definitely does that, I worry that it may run into #50086-ish issues with config keys that aren't intended to be proper valid configs at all. Noticing a problem as soon as you tell us that you're using a given key as a connection [but not before] seems ideal, maybe?

kmcphillips replied:
Oooo that's a very good question. It might be! I think you're right. I'll remove the initializer and see.

Unfortunately though, moving the validation up into the pool invalidates what I was trying to fix originally. The build is green, but the pool is configured and validates before the `on_load(:active_record)` defined in app initializers runs. The repro steps above now fail.

I think I'll try to design an integration test that tests this rather than manual reproduction.

But. It sends me back to how to sequence this such that you can register connection adapters before something validates the adapter in initialization but after Active Record is loaded.

kmcphillips commented:
Alright this is ready for review. It's changed quite a bit so here's what's in here now:

* Move connection validation out of `DatabaseConfig#new` and into `ConnectionHandler` right before the pool is initialized.
* Ultimately did not need an initializer to validate the connections, above should be sufficient?
* Added a new hook `:active_record_register_adapters` which fires right after adapters are registered.
* Fixed CI and got better test coverage. Most notable an integration test that tests adding an adapter in an application level initializer.

The big compromise in here is needing the `on_load(:active_record_register_adapters)` hook. The reason is that in order to register a new adapter it needs to be _after_ Active Record is loaded, but _before_ the DB config is loaded and thus validated. So we cannot use `on_load(:active_record)` as the DB config is validated earlier on in the same hook. We need something that can be triggered from an application level initializer (this is what the integration test covers).

kmcphillips commented on railties/lib/rails/commands/dbconsole/dbconsole_command.rb:
```diff
@@ -51,7 +53,7 @@ def environment
     private
       def adapter_class
         ActiveRecord::ConnectionAdapters.resolve(db_config.adapter)
-      rescue LoadError, ActiveRecord::AdapterNotFound
```
This was swallowing the helpful `AdapterNotFound` error message and raising a subsequent `NotImplementedError` on the `AbstractAdapter` it was falling back to.

byroot commented on activerecord/lib/active_record/connection_adapters.rb:
```diff
@@ -13,19 +13,25 @@ class << self
+      #
+      # == Example
+      #
+      #   ActiveSupport.on_load(:active_record_register_adapters) do
```
Do we really need a callback for this? Maybe I missed some discussion?

kmcphillips replied:
I hope not! Let me walk you through how I got here.

Adapters need to be registered _after Active Record is loaded_ since registration is called on `ActiveRecord::ConnectionAdapters`, but _before the connection config is loaded and validated_ from within the `ConnectionHandler`. That comes from the `initialize_database` railtie here:
https://github.com/rails/rails/blob/438cad462638b02210fc48b700c29dcd0428a8b7/activerecord/lib/active_record/railtie.rb#L305-L311

So you can only register adapters after AR is loaded, but you can't use `on_load(:active_record` because the above railtie is run by the framework earlier than any of the same hook that can be registered in the application, and because there is no `prepend:` type sequencing on these hooks, the one that validates the connections will always run first because it is declared first.

This hook would then fire at exactly the right time and in the right context for other adapters to be registered.

The only alternative I can see is to further delay validating the adapter in the connection, but that means we would end up with connection handler pools we know to be invalid, that we are hoping an adapter will be registered for later. I guess we can move back to using an `after_initialize` railtie to explicitly validate them. But I'm not sure that's better.

matthewd replied:
> Adapters need to be registered after Active Record is loaded

While this sounds literally true (the `ActiveRecord` constant needs to be there), I don't think it's actually true in the `on_load` sense. I'm not immediately seeing why anything more than `require "active_record"` would be needed.

kmcphillips replied:
Oh boy, I'm overcomplicating things.

Yes, that appears to work, so long as the app has `require "active_record"` the registration can be called directly in an initializer and called early but not too early. I've updated the integration test to reflect that.

byroot commented:
I merged another PR, and now this one need a rebase. Sorry :/

kmcphillips commented:
Jean how dare you.

Ok, rebased and squashed. Got to a much simpler solution, thank you for the feedback.
